### PR TITLE
[Typing] Mypy typing part 2

### PR DESCRIPTION
**Pull request by rkooo567**
NOTE: There are many fields that are lazy initialized and assume these are accessed only after lazy initialization is done. I fixed them by using the solution suggested in this approach; https://stackoverflow.com/questions/60925137/using-mypy-with-with-lazy-initialization-of-instance-attributes

Handles some parts of https://github.com/vllm-project/vllm/issues/3680

```
mypy vllm/engine/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/worker/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/spec_decode/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/model_executor/*.py --follow-imports=skip --config-file pyproject.toml
```

Remaining:
```
# mypy vllm/lora/*.py --follow-imports=skip --config-file pyproject.toml

And some other nested directories (such as model_executors/layer)
```

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -379,7 +383,8 @@ def _error_callback(self, exc: Exception) -> None:
 
     async def get_tokenizer(self) -> "PreTrainedTokenizer":
         if self.engine_use_ray:
-            return await self.engine.get_tokenizer.remote()
+            breakpoint()
+            return await self.engine.get_tokenizer.remote()  # type: ignore
```
Comment (rkooo567) [inline vllm/engine/async_llm_engine.py:387]: all these engine related code seems a bit hacky and it was difficult to fix

File: vllm/model_executor/model_loader.py [hunk 2]
```diff
@@ -116,7 +116,7 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
                 # to retain tensorizer args from CLI.
                 model_config.load_format = ParameterizedLoadFormat(
                     model_config.load_format)
-                model_config.load_format.params = (
+                model_config.load_format.params = (  # type: ignore
```
Comment (rkooo567) [inline vllm/model_executor/model_loader.py:119]: @sangstar I found this approach is a bit hacky (dynamically loading load_format) and doesn't work well with typing. Is there any good suggestion to fix it?

**Comment by rkooo567** [comment]
cc @simon-mo @zhuohan123 this PR is ready to review

**Review by simon-mo — APPROVED** [review]
Thanks for working on this. I made a quick skim.

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 3]
```diff
@@ -21,14 +21,62 @@
...
 from transformers import PreTrainedTokenizerBase
 
 
 class BaseLogitsProcessor:
 
+    def __init__(self):
+        # Child class should use initialize in their init.
+        self.fsm: FSM
+
+    def adapt_tokenizer(self, tokenizer: PreTrainedTokenizerBase):
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:36]: yep bad merge. fixed!

**Comment by rkooo567** [comment]
@simon-mo I think we can merge it (I believe spec decoding failure is from the master)

File: vllm/model_executor/model_loader.py (hunk 2 above)
Comment (sangstar) [reply vllm/model_executor/model_loader.py]: I believe @Yard1 's refactor addressed this :D
